### PR TITLE
tests: Revise `process_check_restart.bats`

### PR DESCRIPTION
**Pull request by polarathene**
# Description

General improvements as noted below, makes the test more robust for any future additions.

**Revisions:**
- Removed some outdated comments about wrappers / timings that are no longer relevant.
- `pgrep` check now uses `--full` which was [initially required for MTA-STS support to pass the test](https://github.com/docker-mailserver/docker-mailserver/pull/3592#pullrequestreview-1814634068), whilst now it is more useful for matching postfix/amavis better.
- `pkill` now kills the process and asserts against the oldest matched process PID from `pgrep`. This should be the process that spawned any relevant children. Matching against the PID is better for targeting what we're interested in, since the process path is not always consistent (_`/usr/sbin/amavi` vs `/usr/sbin/amavisd-new`_).
- Added context for the python3 and postgrey services, along with additional maintainance comments on `pkill` / `pgrep` usage.

## Type of change

- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] **I have added information about changes made in this PR to `CHANGELOG.md`**

### CODE REVIEW

File: test/tests/parallel/set3/container_configuration/process_check_restart.bats [hunk 1]
```diff
@@ -153,14 +149,18 @@ function _should_restart_when_killed() {
...
   assert_output --partial "${PROCESS}"
 
   # Should kill the process successfully:
   # (which should then get restarted by supervisord)
-  _run_in_container pkill --echo "${PROCESS}"
-  assert_output --partial "${PROCESS}"
+  # NOTE: The process name from `pkill --echo` does not always match the equivalent processs name from `pgrep --list-full`.
+  # The oldest process returned (if multiple) should be the top-level process launched by supervisord,
+  # the PID will verify the target process was killed correctly:
+  local PID=$(docker exec "${CONTAINER_NAME}" pgrep --full --oldest "${PROCESS}")
```
Comment (georglauterbach) [inline test/tests/parallel/set3/container_configuration/process_check_restart.bats:161]:
```suggestion
  local PID=$(_exec_in_container pgrep --full --oldest "${PROCESS}")
```

File: test/tests/parallel/set3/container_configuration/process_check_restart.bats [hunk 2]
```diff
@@ -176,13 +176,16 @@ function _should_restart_when_killed() {
...
 function _check_if_process_is_running() {
   local PROCESS=${1}
   local MIN_SECS_RUNNING
   [[ -n ${2:-} ]] && MIN_SECS_RUNNING=('--older' "${2}")
 
-  local IS_RUNNING=$(docker exec "${CONTAINER_NAME}" pgrep --list-full "${MIN_SECS_RUNNING[@]}" "${PROCESS}")
+  # `--list-full` provides information for matching against (full process path)
+  # `--full` allows matching the process against the full path (required if a process is not the exec command, such as started by python3 command without a shebang)
+  # `--oldest` should select the parent process when there are multiple results, typically the command defined in `supervisor-app.conf`
+  local IS_RUNNING=$(docker exec "${CONTAINER_NAME}" pgrep --full --list-full "${MIN_SECS_RUNNING[@]}" --oldest "${PROCESS}")
```
Comment (georglauterbach) [inline test/tests/parallel/set3/container_configuration/process_check_restart.bats:188]:
```suggestion
  local IS_RUNNING=$(_exec_in_container pgrep --full --list-full "${MIN_SECS_RUNNING[@]}" --oldest "${PROCESS}")
```

**Review by georglauterbach — CHANGES_REQUESTED** [review]
Just two smaller nitpicks; other than that LGTM 👍🏼

**Comment by polarathene** [comment]
Weird test failure unrelated to changes in this PR 🤔 

```
not ok 71 [setup.sh] show usage when no arguments provided in 131ms
# (from function `assert_success' in file test/test_helper/bats-assert/src/assert_success.bash, line 42,
#  in test file test/tests/parallel/set3/scripts/setup_cli.bats, line 30)
#   `assert_success' failed
#
# -- command failed --
# status : 126
# output : ttrpc: closed: unknown
# --
#
```

https://github.com/docker-mailserver/docker-mailserver/blob/2bf52342500726e86362ee1ea5954b7d96533d32/test/tests/parallel/set3/scripts/setup_cli.bats#L28-L32

---

A quick search on `ttrpc: closed: unknown` shows it's related to running a container, but the results are fairly old. Perhaps something went awry on the CI instance 🤷‍♂️ I'll run it again after and it should pass.

**Review by georglauterbach — APPROVED** [review]

**Pull request merged**